File: common/src/main/java/com/ssblur/scriptor/data_components/BookOfBooksData.java
package com.ssblur.scriptor.data_components;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.item.ItemStack;

import java.util.List;

public record BookOfBooksData(List<ItemStack> items, int active) {
  public static final Codec<BookOfBooksData> CODEC = RecordCodecBuilder.create(instance -> instance.group(
    ItemStack.CODEC.listOf().fieldOf("items").forGetter(data -> data.items),
    Codec.INT.fieldOf("active").forGetter(data -> data.active)
  ).apply(instance, BookOfBooksData::new));
  public static final StreamCodec<RegistryFriendlyByteBuf, BookOfBooksData> STREAM_CODEC = StreamCodec.composite(
    ItemStack.STREAM_CODEC.apply(ByteBufCodecs.list()), BookOfBooksData::items,
    ByteBufCodecs.VAR_INT, BookOfBooksData::active,
    BookOfBooksData::new
  );
}

File: common/src/main/kotlin/com/ssblur/scriptor/data/components/BookOfBooksData.kt
package com.ssblur.scriptor.data.components

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.world.item.ItemStack

@JvmRecord
data class BookOfBooksData(val items: List<ItemStack>, val active: Int) {
  companion object {
    val CODEC: Codec<BookOfBooksData> =
      RecordCodecBuilder.create { instance: RecordCodecBuilder.Instance<BookOfBooksData> ->
        instance.group(
          ItemStack.CODEC.listOf().fieldOf("items").forGetter { data: BookOfBooksData -> data.items },
          Codec.INT.fieldOf("active").forGetter { data: BookOfBooksData -> data.active }
        ).apply(instance) { items: List<ItemStack>, active: Int -> BookOfBooksData(items, active) }
      }
    val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf, BookOfBooksData> = StreamCodec.composite(
      ItemStack.STREAM_CODEC.apply(ByteBufCodecs.list()), BookOfBooksData::items,
      ByteBufCodecs.VAR_INT, BookOfBooksData::active
    ) { items: List<ItemStack>, active: Int -> BookOfBooksData(items, active) }
  }

  override fun equals(other: Any?): Boolean {
    if(other !is BookOfBooksData) return false
    if(other.items.zip(items).any{ it.first != it.second }) return false
    return true
  }

  override fun hashCode(): Int = items.sumOf { it.hashCode() }
}
